ui/src/shared/apis/queryCache.ts

```diff
@@ -0,0 +1,194 @@
+  let hash = 0
+  if (!rawText) {
+    return `${hash}`
+  }
+  for (let i = 0; i < rawText.length; i++) {
+    hash = (hash << 5) - hash + rawText.charCodeAt(i)
+    hash |= 0 // Convert to 32bit integer
+  }
+  return `${hash}`
+}
+
+class QueryCache {
+  cache = {}
+
+  private cleanExpiredQueries = (): void => {
+    const now = Date.now()
+    for (const id in this.cache) {
+      // TODO(ariel): need to implement specific rules for custom time ranges
+      if (this.cache[id].isCustomTime) {
+        continue
+      }
+      if (now - this.cache[id].dateSet > TIME_INVALIDATION) {
+        this.resetCacheByID(id)
+      }
+    }
+  }
+
+  getFromCache = (
+    id: string,
+    hashedVariables: string
+  ): RunQueryResult | null => {
+    // no existing query match
+    if (!this.cache[id]) {
+      return null
+    }
+    // query match with no existing variable match
+    if (this.cache[id].hashedVariables !== hashedVariables) {
+      this.resetCacheByID(id)
+      return null
+    }
+    // query has been initialized but the result has not been set
+    if (this.cache[id].values === undefined) {
+      return null
+    }
+    // query & variable match with an expired result
+    if (Date.now() - this.cache[id].dateSet > TIME_INVALIDATION) {
+      this.resetCacheByID(id)
+      return null
+    }
+    event('query_cache_hit', {context: 'queryCache'})
+    return this.cache[id].values
+  }
+
+  initializeCacheByID = (
+    queryID: string,
+    hashedVariables: string,
+    isCustomTime: boolean = false
+  ) => {
+    if (this.cache[queryID]) {
+      return this.cache[queryID]
+    }
+    this.cache[queryID] = {
+      dateSet: Date.now(),
+      hashedVariables,
+      isCustomTime,
+      mutex: RunQueryPromiseMutex<RunQueryResult>(),
+    }
+    return this.cache[queryID]
+  }
+
+  resetCacheByID = (id: string): void => {
+    if (!this.cache[id]) {
+      return
+    }
+    delete this.cache[id]
+  }
+
+  resetCache = (): void => {
+    this.cache = {}
+  }
+
+  setCacheByID = (
+    queryID: string,
+    hashedVariables: string,
+    values: RunQueryResult
+  ): void => {
+    event('query_cache_miss', {context: 'queryCache'})
+    this.cache[queryID] = {
+      ...this.initializeCacheByID(queryID, hashedVariables),
+      dateSet: Date.now(),
+      values,
+    }
+  }
+
+  startWatchDog = () => {
+    setInterval(() => {
+      this.cleanExpiredQueries()
+    }, TIME_INVALIDATION / 2)
+
+    this.cleanExpiredQueries()
+  }
+}
+
+const queryCache = new QueryCache()
+// Set an interval to check for expired data to invalidate
+queryCache.startWatchDog()
+
+export const resetQueryCache = (): void => {
+  queryCache.resetCache()
+}
+
+export const resetQueryCacheByQuery = (query: string): void => {
+  const queryID = `${hashCode(query)}`
+  queryCache.resetCacheByID(queryID)
+}
+
+export const getCachedResultsOrRunQuery = (
+  orgID: string,
+  query: string,
+  state: AppState
+): CancelBox<RunQueryResult> => {
```

Comment:
throw an event up here too so that we can split questions like "how many queries did the cache save?" and it'd be `queries requested` vs `query cache set`

Reply:
i think this should be taken care of already in the sense that we already report events for `runQuery` which this inevitably calls

Reply:
i dont understand this. feels like an gap between intent vs implementation. explicitly defining the event at the intent (as opposed to using a proxy event rooted in implementation to estimate the intent) helps keep a consistent history to the event, even through implementation changes. if we rewrite all of executeQueries, and forget to put in the eventing, or change the definition of the submodule's eventing structure, our dashboards for cache hits vs missus shouldn't change right? isn't the event you're planning on proxying already invalid because we don't use executeQueries for dashboards?

Reply:
You're right in that the TimeSeries component does not depend on the executeQueries thunk to run a query, it does, however, a duplicate event. I do see what you're saying though. I'll add it in